src/main/java/org/openstreetmap/atlas/checks/validation/tag/TollValidationCheck.java

```diff
@@ -0,0 +1,624 @@
+    }
+
+    @Override
+    protected List<String> getFallbackInstructions()
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    /**
+     * @param edge1
+     *            just an edge
+     * @param edge2
+     *            just another edge
+     * @return angle in degrees between edges (segments)
+     */
+    private Angle angleBetweenEdges(final Edge edge1, final Edge edge2)
+    {
+        final Optional<Heading> edge1heading = edge1.asPolyLine().finalHeading();
+        final Optional<Heading> edge2heading = edge2.asPolyLine().initialHeading();
+        if (edge1heading.isPresent() && edge2heading.isPresent())
+        {
+            return edge1heading.get().difference(edge2heading.get());
+        }
+        return Angle.NONE;
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if the barrier tag contains 'toll'
+     */
+    private boolean barrierTagContainsToll(final Map<String, String> tags)
+    {
+        return tags.get(BarrierTag.KEY).contains(TollTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some edge tags
+     * @param tags2
+     *            some other edge tags
+     * @return both sets of tags have toll=yes
+     */
+    private boolean bothTollYesTag(final Map<String, String> tags, final Map<String, String> tags2)
+    {
+        return this.hasTollYesTag(tags) && this.hasTollYesTag(tags2);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'barrier'
+     */
+    private boolean containsBarrierTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(BarrierTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'highway'
+     */
+    private boolean containsHighwayTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(HighwayTag.KEY);
+    }
+
+    /**
+     * @param tags
+     *            some osm tags
+     * @return boolean if tags contains key 'toll'
+     */
+    private boolean containsTollTag(final Map<String, String> tags)
+    {
+        return tags.containsKey(TollTag.KEY);
+    }
+
+    /**
+     * @param edge
+     *            some edge
+     * @return boolean if edge intersects toll feature
+     */
+    private boolean edgeIntersectsTollFeature(final Edge edge)
+    {
+        final Iterable<Area> intersectingAreas = edge.getAtlas().areasIntersecting(edge.bounds());
+        final Iterable<Node> edgeNodes = edge.connectedNodes();
+        for (final Area area : intersectingAreas)
+        {
+            final boolean areaContainsPolyline = area.asPolygon().overlaps(edge.asPolyLine());
+            final Map<String, String> areaTags = area.getOsmTags();
+            if (!areaContainsPolyline)
+            {
+                continue;
+            }
+            return this.containsBarrierTag(areaTags) && this.barrierTagContainsToll(areaTags);
```

Comment:
Is it desired behavior to short circuit here with possibly returning false and not checking other areas/nodes?

Reply:
"continue" won't short circuit the whole function, it will only move to the next area in the Iterable for loop. Nodes will only be checked if and only if there are no areas intersecting the edge. False will only be returned if there are no areas or nodes intersecting the edge.

Reply:
This is a good point. It shouldn't. I will update this. I think I adjusted this to reduce the cognitive complexity.